### PR TITLE
Handle constants

### PR DESCRIPTION
**Pull request by tserg**
Handle constants 

**What I did**
- For constants that fit into a felt size, exclude them from StorageAllocator and StorageVariableAccessRewriter passes, and print them as Cairo constants.
- For constants that are greater than a felt size (i.e. uint256), add a new pass (ConstantHandler) to inject the constant as a variable into each function that has a reference to it.

### CODE REVIEW

File: src/passes/storageAllocator.ts [hunk 1]
```diff
@@ -26,14 +27,16 @@ export class StorageAllocator extends ASTMapper {
...
     node.vStateVariables.forEach((v) => {
-      const width = CairoType.fromSol(
-        getNodeType(v, ast.compilerVersion),
-        ast,
-        TypeConversionContext.StorageAllocation,
-      ).width;
-      allocations.set(v, usedStorage);
-      usedStorage += width;
-      extractInitialisation(v, initialisationBlock, ast);
+      if (v.mutability !== Mutability.Constant) {
```
Comment (DomHenderson) [inline src/passes/storageAllocator.ts:30]: I think it would be safer to use `!isCairoConstant` for the condition here, to ensure that no variables are both ignored here and later on

File: src/passes/storage/storageVariableAccessRewriter.ts [hunk 2]
```diff
@@ -56,12 +57,14 @@ export class StorageVariableAccessRewriter extends ASTMapper {
       error('VariableDeclaration.vType should be defined for compiler versions > 0.4.x'),
     );
 
-    const parent = node.parent;
-    const replacementFunc = ast.getUtilFuncGen(node).storage.read.gen(node, decl.vType);
-    ast.replaceNode(node, replacementFunc, parent);
-    // Do not recurse
-    // The only argument to replacementFunc is node,
-    // recursing would cause an infinite loop
+    if (decl.mutability !== Mutability.Constant) {
```
Comment (DomHenderson) [inline src/passes/storage/storageVariableAccessRewriter.ts:60]: Also here, if `isCairoConstant` is the check for which variables get printed as cairo constants, `!isCairoConstant` should be the check for which get processed here

File: src/passes/constantHandler.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import { AST } from '../ast/ast';
+import { ASTMapper } from '../ast/mapper';
+import { collectUnboundVariables } from './loopFunctionaliser/utils';
+import { primitiveTypeToCairo } from '../utils/utils';
+
+export class ConstantHandler extends ASTMapper {
+  // Visit all functions and inject 256 bit constants into the functions that
+  // refer to these constants. Constants less than 256 bits are handled as native
+  // Cairo constants that are directly printed in cairoWriter.ts
+  // Note that all constants and are excluded from the storage passes.
```
Comment (DomHenderson) [inline src/passes/constantHandler.ts:20]: This is either missing a word or has an extra and

**Review by DomHenderson — CHANGES_REQUESTED** [review]
This looks really good. I've just added a few small comments and highlighted one particularly subtle edge case

File: tests/behaviour/expectations/behaviour.ts [hunk 4]
```diff
@@ -986,6 +1052,7 @@ export const expectations = flatten(
           Expect.Simple('test', [], []),
           Expect.Simple('returnTest', [], ['12', '0']),
         ]),
+        */
```
[the remark this answers is not in the document]
Comment (tserg) [reply tests/behaviour/expectations/behaviour.ts:1055]: Sorry about this! I was trying to speed up the testing. Will uncomment them

File: src/transpiler.ts [hunk 5]
```diff
@@ -70,11 +71,12 @@ function applyPasses(ast: AST, options: TranspilationOptions): AST {
   const passes: Map<string, typeof ASTMapper> = createPassMap([
     ['Ss', SourceUnitSplitter],
     ['Ru', RejectUnsupportedFeatures],
+    ['Ib', IntBoundCalculator],
```
[the remark this answers is not in the document]
Comment (tserg) [reply src/transpiler.ts:74]: I am playing around with this example - even though the contract compiles, all three functions revert on runtime. Should we replicate this behaviour? E.g. check if the value exceeds the bounds of the constant's type, insert a failing assertion before each reference.

```
pragma solidity ^0.8.6;

// SPDX-License-Identifier: MIT

contract Warp {
  int16 x = type(int16).max -10000;
  int16 constant y = type(int16).max + 1;

  function getY() public pure returns (int16) {
    return y;
  }

  function getZ() public view returns (int16) {
    return x - y;
  }

  function pureY() public pure returns (int16) {
    return y + 1;
  }
}
```

[the remark this answers is not in the document]
Comment (DomHenderson) [reply src/transpiler.ts:74]: Good catch. It does bring up a nasty issue of constants that work in unchecked blocks, but cause a failing assertion if used outside of them. There's a few other concerns I have that could arise from using the IntBoundCalculator before the LiteralExpressionEvaluator, but I think for this PR constants like those in this example should be treated like constants assigned the value returned from a function and kept as state variables for now, and another PR can potentially be opened to handle this case more efficiently
